### ISSUE TITLE
Tagless Final Encoding

### ISSUE THREAD

**Issue by vlovgr** [vlovgr/ciris#39]
The idea is to parametrise `ConfigSource` with an additional type parameter `F[_]`, so that the type signature becomes `ConfigSource[F[_], Key]`, and so that it now reads keys with return type `F[ConfigSourceEntry[Key]]`.

Remaining parts of Ciris, including documentation, will need to be updated to account for the type parameterisation. Existing functions can use the `Id` type as context. New functions will need to be created where the context can be specified or inferred.

This would allow to write code similar to the following, replacing `Future` with some other type for which you can define a `Monad` instance.

```scala
type F[A] = Future[A]

val source: ConfigSource[F, String] = ???

def fetch[Value: ConfigReader](key: String): F[ConfigValue[Value]] = ???

val sum: F[BigDecimal] =
  loadConfig(
    fetch[BigDecimal]("FIRST_KEY"),
    fetch[Int]("SECOND_KEY"),
    fetch[Double]("THIRD_KEY")
  )(_ + _ + _)
```

This work will also involve:
- Replicate parts of `Id`, `Applicative`, and `Monad` from Cats in the core module. 
  This is to ensure we stay dependency-free in the core module.
- Provide conversions from Cats' instances to the core module's instances.
   These conversions should fit within the same Cats module described in #38.

**Comment by vlovgr** [comment]
The interest/need for this seems low, and it significantly increases the complexity of the library, so closing this for now.

**Comment by vlovgr** [comment]
Fixed in #102.
